compiler/rustc_lexer/src/unescape.rs

```diff
@@ -80,203 +81,253 @@ impl EscapeError {
     }
 }
 
-/// Takes the contents of a unicode-only (non-mixed-utf8) literal (without
-/// quotes) and produces a sequence of escaped characters or errors.
-///
-/// Values are returned by invoking `callback`. For `Char` and `Byte` modes,
-/// the callback will be called exactly once.
-pub fn unescape_unicode<F>(src: &str, mode: Mode, callback: &mut F)
-where
-    F: FnMut(Range<usize>, Result<char, EscapeError>),
-{
-    match mode {
-        Char | Byte => {
-            let mut chars = src.chars();
-            let res = unescape_char_or_byte(&mut chars, mode);
-            callback(0..(src.len() - chars.as_str().len()), res);
-        }
-        Str | ByteStr => unescape_non_raw_common(src, mode, callback),
-        RawStr | RawByteStr => check_raw_common(src, mode, callback),
-        RawCStr => check_raw_common(src, mode, &mut |r, mut result| {
-            if let Ok('\0') = result {
-                result = Err(EscapeError::NulInCStr);
-            }
-            callback(r, result)
-        }),
-        CStr => unreachable!(),
-    }
-}
-
 /// Used for mixed utf8 string literals, i.e. those that allow both unicode
 /// chars and high bytes.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
 pub enum MixedUnit {
-    /// Used for ASCII chars (written directly or via `\x00`..`\x7f` escapes)
+    /// Used for ASCII chars (written directly or via `\x01`..`\x7f` escapes)
     /// and Unicode chars (written directly or via `\u` escapes).
     ///
     /// For example, if '¥' appears in a string it is represented here as
     /// `MixedUnit::Char('¥')`, and it will be appended to the relevant byte
     /// string as the two-byte UTF-8 sequence `[0xc2, 0xa5]`
-    Char(char),
+    Char(NonZero<char>),
```

Comment:
UTF-8 includes the value 0x00, which may be written in a Rust string like `\0`. I do not see how we can accept this PR.

Reply:
*headscratch* Stared at things a bit longer. If it is only used in the case of `c""`, then the documentation of this type should be changed to reflect the fact that it is an implementation detail of Specifically That, as this is misleading as-is.

Reply:
My understanding however was that there was a desire to eventually have "mixed strings which are not necessarily CStr", which is probably why the type is described in a more generic way and doesn't reference CStr currently.

Reply:
It's currently only used for non-raw C strings AFAIK as raw C strings are perfectly happy with char.
@nnethercote's https://github.com/rust-lang/rust/commit/a1c07214f0f7988cbc5a645a499bb8f7dd9cbed7 does mention that "it will soon be used in more than just C string literals.", but I don't know what that refers to. Seems it's for https://rust-lang.github.io/rfcs/3349-mixed-utf8-literals.html which would mean byte strings would also use this construction and those are fine with null bytes. So either we don't use NonZero for C strings or we use a different unit-type for C strings and for byte strings.